oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/query/IoTDBBrowserLogQueryDAO.java

```diff
@@ -0,0 +1,126 @@
+        }
+        if (Objects.nonNull(category)) {
+            query.append(" and ").append(BrowserErrorLogRecord.ERROR_CATEGORY).append(" = ").append(category.getValue());
+        }
+        query.append(IoTDBClient.ALIGN_BY_DEVICE);
+        // IoTDB doesn't support the query contains "1=1" and "*" at the meantime.
+        String queryString = query.toString();
+        queryString = queryString.replace("1=1 and ", "");
+
+        List<? super StorageData> storageDataList = client.filterQuery(BrowserErrorLogRecord.INDEX_NAME, queryString, storageBuilder);
+        List<BrowserErrorLogRecord> browserErrorLogRecordList = new ArrayList<>(storageDataList.size());
+        storageDataList.forEach(storageData -> browserErrorLogRecordList.add((BrowserErrorLogRecord) storageData));
+        // resort by self, because of the select query result order by time.
+        browserErrorLogRecordList.sort((BrowserErrorLogRecord b1, BrowserErrorLogRecord b2) -> Long.compare(b2.getTimestamp(), b1.getTimestamp()));
+        BrowserErrorLogs logs = new BrowserErrorLogs();
+        int limitCount = 0;
+        for (int i = 0; i < browserErrorLogRecordList.size(); i++) {
+            if (i >= from && limitCount < limit) {
+                limitCount++;
+                BrowserErrorLogRecord record = browserErrorLogRecordList.get(i);
+                if (CollectionUtils.isNotEmpty(record.getDataBinary())) {
+                    BrowserErrorLog log = iotdbParserDataBinary(record.getDataBinary());
+                    logs.getLogs().add(log);
+                }
+            }
+        }
+        logs.setTotal(storageDataList.size());
+        return logs;
+    }
+
+    private BrowserErrorLog iotdbParserDataBinary(byte[] dataBinaryBase64) {
+        try {
+            BrowserErrorLog log = new BrowserErrorLog();
+            org.apache.skywalking.apm.network.language.agent.v3.BrowserErrorLog browserErrorLog = org.apache.skywalking.apm.network.language.agent.v3.BrowserErrorLog
+                    .parseFrom(dataBinaryBase64);
+            log.setService(browserErrorLog.getService());
+            log.setServiceVersion(browserErrorLog.getServiceVersion());
+            log.setTime(browserErrorLog.getTime());
+            log.setPagePath(browserErrorLog.getPagePath());
+            log.setCategory(ErrorCategory.valueOf(browserErrorLog.getCategory().name().toUpperCase()));
+            log.setGrade(browserErrorLog.getGrade());
+            log.setMessage(browserErrorLog.getMessage());
+            log.setLine(browserErrorLog.getLine());
+            log.setCol(browserErrorLog.getCol());
+            log.setStack(browserErrorLog.getStack());
+            log.setErrorUrl(browserErrorLog.getErrorUrl());
+            log.setFirstReportedError(browserErrorLog.getFirstReportedError());
+
+            return log;
+        } catch (InvalidProtocolBufferException e) {
```

Comment:
Is this an IOException? If so, we don't need this try/catch.

Reply:
No. This is method is refer to `IBrowserLogQueryDAO.parserDataBinary` without Base64 decode, because it has been decoded in `BrowserErrorLogRecord.Builder.storage2Entity`. The InvalidProtocolBufferException is thrown by `org.apache.skywalking.apm.network.language.agent.v3.BrowserErrorLog.parseFrom`